### ISSUE TITLE
Implement SetSpecifier

### ISSUE THREAD

**Issue by k-rus** [Starcounter/Starcounter#2831]
To be able to add retrievable records set specifier needs to be allocated. For this table SetSpecifier is created, which maps types and already allocated set specifier. A set specifier or TypeId is a string with a special format. See [this comment](https://github.com/Starcounter/Starcounter/issues/2160#issuecomment-127505943) for formal specification and [this comment](https://github.com/Starcounter/Starcounter/issues/2160#issuecomment-65769307) for high level description.

The format of SetSpecifier table is simple:
```xml
		"TableName" : "SetSpecifier",
		"Columns" : [
			{
				"ColumnName" : "TypeId",
				"TypeCode" : "STAR_TYPE_STRING"
			},
			{
				"ColumnName" : "Type",
				"TypeCode" : "STAR_TYPE_REFERENCE"
			}
		],
		"Indexes" : [
			{
				"IndexName" : "TypeId",
				"Unique" : true,
				"ColumnNames" : ["TypeId"]
			},
			{
				"IndexName" : "Type",
				"Unique" : true,
				"ColumnNames" : ["Type"]
			}
		]
```
Column `TypeId` stores allocated set specifier strings. Column `Type` contains records/objects, which describe types and can be inserted after the most meta-tables were created. Currently, they belong to table `RawView`, which extends `Table`, which extends `Type`.

Adding a record to `SetSpecifier` will allocate set specifier with a next letter in the collation on the corresponding inheritance level. For example, a new table extends existing table, which has a set specifier `~a` and has already one child with a set specifier `~a~a`. Thus the new table gets set specifier `~a~b`. If it was the last letter in the encoding, i.e., `_`, then first two letters, i.e., `aa` is allocated.

SetSpecifier string values are needed to insert new records and to create indexes. There are several needs during preparation steps:

1. Insert new tokens and access existing tokens, i.e., records in table `Token`. This is a predefined set specifier: `#define STAR_SETSPEC_TOKEN u"~a"`
2. Create indexes, which requires to provide set specifier.
3. Generate API to search and insert meta-data.

I see three approaches to create set specifiers for needs 2 and 3:

1. Predefine set specifier during API generation. Insert  type and set specifier records afterwords. This requires to implement allocation algorithm for the predefine in addition to the algorithm for inserting into set specifier table. Note that the inserting afterwords should avoid the insertion algorithm.
2. Allocate set specifiers by inserting into SetSpecifier without providing type records. Set type records afterwords after the records were created. This requires to predefine set specifier for table/type `SetSpecifier`.
3. Create type records first and then allocate set specifier. This requires to create the tables first, allocate setspecifier and then to create indexes. This also requires to predefine quite many set specifiers.

Just a side note: SetSpecifier contains reference to some record, since records/obejcts can be referred as type without extending any known table and without name in future.

**Comment by k-rus** [comment]
> If it was the last letter in the encoding, i.e., _, then first two letters, i.e., aa is allocated.

My definition on the next and last letter is not correct. The algorithm does not need to maintain the same order as the collation order. The only important thing is that all letters of the collation are utilised before increasing number of letters.

**Comment by alexanderstc** [comment]
I think the delimiter (~) "position" or "precedence" or "sort priority" relative to other characters is different in the setspec collation versus index search collation, which is Unicode-like, if I'm not mistaken.
Not sure if this can interfere with our setspe allocation rules.
In any case, I have a workaround in mind.

**Comment by k-rus** [comment]
> Not sure if this can interfere with our setspec allocation rules.

I think the same. Only the delimiter should require special attention, since we don't not which collation is used for the database and we don't control collation on our column `TypeId` of `SetSpecifier`.

**Comment by alexanderstc** [comment]
A working prototype on the branch CleanMetalayer-alexander. Still need to implement Insert base table's fields in the metalayer generator; make setspec creation code use letters other than a-z.

**Comment by alexanderstc** [comment]
There is some impression of slowness during setspec creation, or probably during meta-table creation, has to be investigated.

**Comment by alexanderstc** [comment]
Implemented (see #2850), closed.
